pkg/lib/serverTime.js

```diff
@@ -321,13 +324,181 @@ export function ServerTime() {
                     console.warn("failed to load time servers", error);
                     return result;
                 });
-    };
+    }
 
-    self.set_custom_ntp = function (servers, enabled) {
-        const text = `# This file is automatically generated by Cockpit\n\n[Time]\n${enabled ? "" : "#"}NTP=${servers.join(" ")}\n`;
+    function set_ntp_config_timesyncd(config) {
+        const custom_ntp_config_file = cockpit.file("/etc/systemd/timesyncd.conf.d/50-cockpit.conf",
+                                                    { superuser: "try" });
+
+        const text = `# This file is automatically generated by Cockpit\n\n[Time]\n${config.enabled ? "" : "#"}NTP=${config.servers.join(" ")}\n`;
 
         return cockpit.spawn(["mkdir", "-p", "/etc/systemd/timesyncd.conf.d"], { superuser: "try" })
                 .then(() => custom_ntp_config_file.replace(text));
+    }
+
+    function find_file_and_read(names, first) {
+        if (first === undefined)
+            first = 0;
+        if (first >= names.length)
+            return Promise.reject(new Error("Found none of " + names.join(",")));
+        return cockpit.file(names[first], { superuser: "try" }).read()
+                .then(text => {
+                    if (text)
+                        return [names[first], text];
+                    return find_file_and_read(names, first + 1);
+                });
+    }
+
+    function get_ntp_config_chronyd() {
+        return find_file_and_read(["/etc/chrony.conf", "/etc/chrony/chrony.conf"])
+                .then(([name, text]) => {
+                    const result = {
+                        supported: true,
+                        backend: "chronyd",
+                        file: name,
+                        standard: [],
+                        sources: [],
+                        extra: []
+                    };
+
+                    let in_cockpit_section = false;
+                    for (const line of text.split("\n")) {
+                        if (line == "# Added by Cockpit") {
+                            in_cockpit_section = true;
```

Comment:
Writing such a comment is okay as a hint for the admin, but the parser must not depend on this.  Admins can add time servers through other means, delete that comment, or even move it around. Again: Reading the config must not be specific to cockpit, with magic syntax. Please just read all the pool/sourcedir/server lines, and give up if you encounter anything that you cannot interpret.

Reply:
> Writing such a comment is okay as a hint for the admin, but the parser must not depend on this.

I don't share this strong conviction; admins don't do random things.

Reply:
They don't, but this is IMHO fundamentally the wrong approach. The criterion should be "cockpit's parser can understand a config line or not", not "somewhere else in the file is some magic marker". What do you do if the "Added by cockpit" section was modified with an option that cockpit doesn't understand? And conversely, if something else added custom time servers, why not show them in cockpit? this seems to just unnecessarily make our life hard.

Reply:
My main goal is to allow people to _add_ more NTP servers to the configuration and get chronyd to actually use them instead of the default ones. A secondary goal is to allow the user to easily revert to the standard configuration. This mirrors what we have for systemd-timesyncd.

If we don't make a distinction between "standard servers" and "servers added by cockpit", then we need to allow the user to remove lines from chrony.conf with Cockpit that have been put there by the admin or the distribution. I am hesitant about this because Cockpit (currently) does not allow the user to put them back, since Cockpit can only add "server" lines without options. If you remove "pool fedora.pool.ntp.org iburst" with Cockpit, you can never put it back with Cockpit.

I still think the current UI is a good compromise.

Reply:
> My main goal is to allow people to add more NTP servers to the configuration and get chronyd to actually use them instead of the default ones.

OK -- but then I think this would be better:

 * cockpit only ever edits/adds arbitrary new servers to /etc/chrony/sources.d/cockpit.conf. That directory is configured by default in Debian already. Let's file a Fedora/RH bug against chrony to add that to the package as well, to establish a standard directory, instead of admins editing the main config file (which will henceforth become un-upgradeable by the distro).
 * Until that gets implemented, I'm willing to accept a hack that will create that directory and add it to the main chrony.conf if it doesn't exist yet. That would never run on Debian, it would run at most once on Fedora/RHEL, and we would never have to remove it.

That's what I meant [earlier on](https://github.com/cockpit-project/cockpit/pull/17992#discussion_r1063979856). My rejection was about /etc/cockpit-chrony-sources/ as namespace trampling, *not* about having a separate sources.d/ directory. Sorry if that was unclear!

But with either approaches it's still not clear to me how you would want to implement "instead of the default ones". If cockpit only ever adds/removes to sources.d/cockpit.conf, or your current magic "## cockpit section", then you can only comment out all other `server`s wholesale. Right now you even just comment out `sourcedir`s, but (1) Fedora/RHEL does not have that , and (2) with a generic sources.d/ it's too coarse-scaled, it would have to be per file.

In summary, I still think there is no design proposal that fits your goals and is also robust with chrony's existing and potential configuration structure. :cry:

Reply:
Alright, this is less ambitious and actually very close to the first iteration.
We just should be honest in the UI and say either "Additional servers that are used when the default servers can't be contacted", or "Additional servers that are tried before the default ones", whichever is the case.

Some people might just need to add a server because the default ones don't work, some people might really want to never even try to contact the default ones. I guess.